beep/utils/waveform.py

```diff
@@ -56,3 +58,223 @@ def convert_velocity_to_power_waveform(waveform_file, velocity_units):
+
+    def get_input_current_multistep_soc_as_x(self, charging_c_rates):
+        """
+        Helper function to generate a waveform with stepwise charging currents. Should return a vector
+        of current values and vector of corresponding state of charge (soc) values.
+
+        Args:
+            charging_c_rates (list): c-rates for each of the charging steps. Each step is assumed to be an equal
+                SOC portion of the charge, and the length of the list just needs to be at least 1
+
+        returns
+        np.array: array with the current as a function of soc
+        np.array: array with the corresponding soc values
+        """
+        mesh_points = np.linspace(self.soc_i, self.soc_f, len(charging_c_rates) + 1)
+        soc_vector = np.linspace(self.soc_i, self.soc_f, self.soc_points)
+        c_rate_i = np.zeros((len(mesh_points) - 1, len(soc_vector)))
+
+        for i, elem in enumerate(mesh_points[0:-1]):
+            soc1 = mesh_points[i]
+            soc2 = mesh_points[i + 1]
+            c_rate_i[i] = charging_c_rates[i] * np.heaviside(soc_vector - soc1, 0.5) * \
+                np.heaviside(soc2 - soc_vector, 0.5)
+
+            multistep_current_raw = sum(c_rate_i)
+
+        c_rate_end = self.final_c_rate * np.heaviside(soc_vector - soc2, 0.5) * \
+            np.heaviside(self.soc_f - soc_vector, 0.5)
+
+        multistep_current = multistep_current_raw + c_rate_end
+
+        return multistep_current, soc_vector
+
+    def get_input_current_smooth_soc_as_x(self, charging_c_rates):
+        """
+        Helper function to generate a waveform with smoothly varying charging current. Should return a vector
+        of current values and vector of corresponding state of charge (soc) values.
+
+        Args:
+            charging_c_rates (list): c-rates for each of the charging steps. Each step is assumed to be an equal
+                SOC portion of the charge, and the length of the list just needs to be at least 1
+
+        returns
+        np.array: array with the current as a function of soc
+        np.array: array with the corresponding soc values
+        """
+        soc_vector = np.linspace(self.soc_i, self.soc_f, self.soc_points)
+        mesh_points = np.linspace(self.soc_i, self.soc_f, len(charging_c_rates) + 1)
+        mesh_points_mid = mesh_points
+        mesh_points_mid[0:-1] += 1 / (len(mesh_points_mid)) * 0.5
+
+        mesh_points_mid = list([self.soc_i]) + list(mesh_points_mid) + list([self.soc_f * 1.01])
+        mesh_points_mid = np.array(mesh_points_mid)
+
+        charging_c_rate_start = self.final_c_rate
+        rates = np.array([charging_c_rate_start] + charging_c_rates + [self.final_c_rate] + [self.final_c_rate])
+        interpolator = interpolate.PchipInterpolator(mesh_points_mid, rates, axis=0, extrapolate=0)
+
+        charging_c_rate_soc1_end = interpolator.__call__(mesh_points[1])
```

Comment:
@pasinger Should this be `mesh_points[-1]`? If so, then some of the smooth curves change substantially when we do the time matching to 80%
![Uploading rapid_charge_matching.png…]()

Reply:
I don't think it should be `mesh_points[-1]`. This value is a part of the routine we had decided on for how to set the C rates in the first SOC window. There was some ambiguity originally because we have the C rates "clamping" at the mid-points of the SOC windows, and it would be unclear what behavior should be set for the first C rate before it's mid-point. Obtaining the C rate value from `mesh_points[1]` is part of the routine to either a) set the slope to constant between the first two SOC points or b) set the C rate before the first SOC mid-point to be CC at its specified value. These are calculated in `charging_c_rate_start` below. Case (a) occurs if the slope will be negative and Case (b) occurs if the slope would have been positive.
It may be worth revisiting this behavior since we have added constraints into the new parameterization.

Reply:
Yeah, I think we can leave the behavior as is for now. I can't really think of a good alternative off-hand.